### ISSUE TITLE
Can bypass `PodSecurityContext.SupplementalGroups` by custom container image although PSP(or other policy engines) enforces the field

### ISSUE THREAD

**Issue by everpeace** [kubernetes/kubernetes#112879]
### What happened?

Even PSP or other policy engines enforces the field values of `PodSecurityContext.SupplementalGroups`, user can bypass the setting by building a custom container image like this.

Assume the below PSP is enforced:

```yaml
kind: PodSecurityPolicy
...
spec:
  # this only allows uid=1000, gid=1000, supplementalGroups=[60000]
  runAsUser:
    ranges: { min: 1000, max: 1000 }
    rule: MustRunAs
  runAsGroup:
    ranges: { min: 1000, max: 1000 }
    rule: MustRunAs
  supplementalGroups:
    ranges: { min: 60000, max: 60000 }
    rule: MustRunAs
```

Then, a user can create a pod that PSP allows with a custom image. But it can bypass `SupplementalGroups` setting by crafting a container image. 

```Dockerfile
# Dockerfile 
# In the image, uid=1000 belongs to 50000
FROM ubuntu:22.04
RUN groupadd -g 50000 bypassed-group \
    && useradd -m -u 1000 alice \
    && gpasswd -a alice bypassed-group
```

```yaml
# Pod
kind: Pod
metadata:
  name: bypass-supplementalgroups-pod
spec:
  # the securityContext satisfies the above PSP
  securityContext:
    runAsUser: 1000
    runAsGroup: 1000
    supplementalGroups: [60000]
  containers:
  # However, because uid=1000 belongs to gid=50000 in the image,
  # the process's identity in the container has groups=50000,60000
  # Note that 50000 is bypassed by container image 
  # even though PSP enforces supplementalGroups=60000
  - image: built-above
    # the command outputs: 'uid=1000(alice) gid=1000(alice) groups=1000(alice),50000(bypassed-group),60000'
    command: ["bash", "-c", "id; true"]
...
```

#### Why does this behavior matter?

In a multi-tenant Kubernetes cluster using `hostPath` volumes, this behavior may be very confusing for cluster administrators. 

`hostPath` volumes are access-controlled by uid/gid as in the usual Linux manner. In such clusters, I think it would be a common practice for the cluster administrator to use PSP or policy engine to restrict the values of `PodSecurityContext.{runAsUser, runAsGroup, SupplementalGroups}` field to secure private directory in `hostPath` volumes.  However, a custom container image can easily bypass the setting and can get access to some private directory in the `hostPath` volume. This behavior confuses cluster administrators. They might say _"What can `supplementalGroups` policy in PSP protect?"_ 

See [here](https://github.com/pfnet-research/strict-supplementalgroups-container-runtime#motivation) for a more detailed and practical impact. 

### What did you expect to happen?

`bypass-supplementalgroups-pod` can have group identities defined only in  `PodSecurityContext.SupplementalGroups`.  That is, the pod's log expects to be

```
uid=1000(alice) gid=1000(alice) groups=1000(alice),60000
```

### How can we reproduce it (as minimally and precisely as possible)?

See https://github.com/pfnet-research/strict-supplementalgroups-container-runtime/tree/reproduce-bypass-supplementalgroups. You can reproduce this behavior in a single command. You can try to reproduce this both with contained and cri-o.

#### Is this a security issue??

We reported this behavior to hackerone.com(`#1688374`) by following [Kubernetes Security and Disclosure Information](https://kubernetes.io/docs/reference/issues-security/security/). Kubernetes Security Response Committee responded that this behavior "works as intended," and they recommended discussing how to handle this behavior in a public k/k issue.

However, I understand this behavior is not recognized widely. So, for clusters using NFS(or similar filesystems) as `hostPath` volumes, I think it's popular for supercomputers or HPC area, this behavior might cause critical consequences because many cluster administrators might understand PSP(or other policy engines) can provide enough protection to private directories in such `hostPath` volumes.

See [here](https://github.com/pfnet-research/strict-supplementalgroups-container-runtime#motivation) for a more detailed and practical impact.

cc/ @cjcullen (as a responder to our report)

#### Root Cause: ambiguous definition

In kubernetes API, [`Pod.SecurityContext.SupplementalGroups`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#podsecuritycontext-v1-core) is defined as _"A list of groups applied to the first process run in each container, in addition to the container's primary GID. If unspecified, no groups will be added to any container. Note that this field cannot be set when spec.os.name is windows."_

I think this definition may be a bit ambiguous for readers regarding the gids defined in the container image. However, most popular CRI implementations(contained and cri-o) create [OCI runtime spec](https://github.com/opencontainers/runtime-spec/blob/main/config.md#user)(`config.json`) with _merged gids_ defined in the container image and `supplementalGroups` in the pod spec like this:

```json5
{
  "ociVersion": "1.0.2-dev",
  "process": {
    "user": {
      "uid": 1000,
      "gid": 1000,
      # CRI merges group gids for uid=1000
      "additionalGids": [
        50000, # defined in container image (uid=1000 belongs to it)
        60000  # defined in securityContext.supplementalGroups
      ]
    },
...
```

### How to resolve this issue?

_added: as described in https://github.com/kubernetes/kubernetes/issues/112879#issuecomment-1276541170, we will improve the API description first, and keep discussing new API in KEP._

First, I would like to discuss how to resolve this in the community. I propose several ways to resolve this.

- Step 1. Improve the description in the API spec and warn this behavior in a blog or similar ways to users
  - Pros
    - Easy. Just Documentation. No breaking changes.
    - This behavior can be mitigated by implementing the custom `RuntimeClass` which enforces it.
      - See [pfnet-research/strict-supplementalgroups-container-runtime](https://github.com/pfnet-research/strict-supplementalgroups-container-runtime)
    - (I'm welcome to write an article for this and its mitigations)
  - Cons
    - This behavior will become official. It might break the "secure by design" principle.
- Step 2. Add a new API: `PodSecurityContext.SupplementalGroupsPolicy={RespectGroupsInImage(default), IgnoreGroupsInImage}` or similar
  - Pros
    - No breaking change.
    - Users can have an official API to select the behavior.
  - Cons
    - Hard. Learge impact.
    - Needs KEP (I'm also welcome to write a KEP)
    - This affects all the CRI implementations
- ~Change an API behavior~
  - ~Pros~
    - ~None.~
  - ~Cons~
    - ~Breaking change.~

/area security
/sig security
/area api
/sig architecture

```[tasklist]
- [x] https://github.com/kubernetes/kubernetes/pull/113047
- [x] https://github.com/kubernetes-sigs/cri-tools/pull/1005
- [ ] https://github.com/kubernetes/enhancements/issues/3619
```

### Kubernetes version

<details>
I checked this behavior in the below version. But I think this might affect all the recent versions.

```console
$ kubectl version
Client Version: v1.24.6
Kustomize Version: v4.5.4
Server Version: v1.25.2
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
PRETTY_NAME="Ubuntu 22.04.1 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.1 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
$ uname -a
Linux reproduce-bypass-supplementalgroups-control-plane 5.10.104-linuxkit #1 SMP PREEMPT Thu Mar 17 17:05:54 UTC 2022 aarch64 aarch64 aarch64 GNU/Linux
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

I tested this behavior both on containerd and cri-o.

<details>
containerd github.com/containerd/containerd v1.6.8 9cd3357b7fd7218e4aec3eae239db1f68a5a6ec6 <br />
crio version 1.25.0
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by everpeace** [comment]
I put labels of sig/security and sig/architecture.  But I'd appreciate it if someone told me more suitable sig or area labels.

**Comment by nitishmalang** [comment]
assign me

**Comment by liggitt** [comment]
routing to sig-node since it owns the Pod API and kubelet / CRI runtime implications

**Comment by everpeace** [comment]
Thanks liggitt for routing this to sig/node!

I look forward to hearing a response from the sig/node.

**Comment by everpeace** [comment]
/assign

We haven't agreed on how to resolve this issue　in the community. However, I would like to contribute to the issue if I can.

**Comment by SergeyKanzhelev** [comment]
Discussed at CI triage meeting. @everpeace will improve documentation of the field as a first step and improvements can be discussed as a KEP.

/remove-kind bug
/kind documentation

**Comment by kfox1111** [comment]
Ouch... yeah, this is unexpected behavior.

**Comment by npolshakova** [comment]
Hi @everpeace,

Checking in as we approach 1.27 code freeze at 17:00 PDT on Tuesday 14th March 2023.

Please ensure the following items are completed:

- [ ] All PRs to the Kubernetes repo that are related to your enhancement are linked in the above issue description (for tracking purposes).
- [ ] All PRs are fully merged by the code freeze deadline.

Please let me know if there are any other PRs in k/k I should be tracking for this KEP.
As always, we are here to help should questions come up. Thanks!

**Comment by everpeace** [comment]
[KEP-3619](https://kep.k8s.io/3619) is alpha-released in [v1.31.0](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.31.md#v1310). I'm closing this.
